src/core/execute.c

```diff
@@ -333,6 +335,40 @@ static void log_command_line(Unit *unit, const char *msg, const char *executable
 
 static int exec_context_load_environment(const Unit *unit, const ExecContext *c, char ***l);
 
+static int direct_spawn_fallback(Unit *unit, FILE *f, char *log_level, FDSet *fdset, pid_t *ret_pid) {
+        char serialization_fd_number[DECIMAL_STR_MAX(int) + 1];
+        int r;
+
+        assert(unit);
+        assert(f);
+        assert(log_level);
+        assert(fdset);
+        assert(ret_pid);
+
+        r = fd_cloexec(fileno(f), false);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to set O_CLOEXEC on serialization fd: %m");
+
+        r = fdset_cloexec(fdset, false);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to set O_CLOEXEC on serialized fds: %m");
+
+        xsprintf(serialization_fd_number, "%i", fileno(f));
+
+        /* The executor binary is pinned, to avoid compatibility problems during upgrades. */
+        r = posix_spawn_wrapper(FORMAT_PROC_FD_PATH(unit->manager->executor_fd),
+                        STRV_MAKE(FORMAT_PROC_FD_PATH(unit->manager->executor_fd),
+                                  "--deserialize", serialization_fd_number,
+                                  "--log-level", log_level,
+                                  "--log-target", log_target_to_string(manager_get_executor_log_target(unit->manager))),
+                        environ,
+                        ret_pid);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to spawn executor: %m");
```

Comment:
can you add a wrapper exec_spawn_executor() or so? you have these very similar posix_spawn_wrapper() invocations multiple times

Reply:
twice, and they use different arguments, as one is the fallback, so a wrapper's wrapper would result in more code and look worse